imports/plugins/included/taxes-taxcloud/server/hooks/hooks.js

```diff
@@ -42,84 +37,85 @@ MethodHooks.after("taxes/calculate", (options) => {
-                  Qty: items.quantity
-                };
-                index += 1;
-                cartItems.push(item);
-              }
+          // format cart items to TaxCloud structure
+          let index = 0;
+          for (const items of cartToCalc.items) {
+            // only processs taxable products
+            if (items.variants.taxable === true) {
+              const item = {
+                Index: index,
+                ItemID: items.variants._id,
+                TIC: "00000",
+                Price: items.variants.price,
+                Qty: items.quantity
+              };
+              index += 1;
+              cartItems.push(item);
             }
+          }
 
-            // request object
-            const request = {
-              headers: {
-                "accept": "application/json",
-                "content-type": "application/json"
-              },
-              data: {
-                apiKey,
-                apiLoginId,
-                customerID: cartToCalc.userId,
-                cartItems,
-                origin,
-                destination,
-                cartID: cartId,
-                deliveredBySeller: false
-              }
-            };
+          // request object
+          const request = {
+            headers: {
+              "accept": "application/json",
+              "content-type": "application/json"
+            },
+            data: {
+              apiKey,
+              apiLoginId,
+              customerID: cartToCalc.userId,
+              cartItems,
+              origin,
+              destination,
+              cartID: cartId,
+              deliveredBySeller: false
+            }
+          };
 
-            HTTP.post(url, request, (error, response) => {
-              let taxRate = 0;
-              // ResponseType 3 is a successful call.
-              if (!error && response.data.ResponseType === 3) {
-                let totalTax = 0;
-                for (const item of response.data.CartItemsResponse) {
-                  totalTax += item.TaxAmount;
-                }
-                // don't run this calculation if there isn't tax.
-                if (totalTax > 0) {
-                  taxRate = (totalTax / cartToCalc.getSubTotal());
-                }
-                // we should consider if we want percentage and dollar
-                // as this is assuming that subTotal actually contains everything
-                // taxable
-                Meteor.call("taxes/setRate", cartId, taxRate, response.CartItemsResponse);
-              } else {
-                let errMsg = "Unable to access service. Check credentials.";
-                if (response && response.data.Messages[0].Message) {
-                  errMsg = response.data.Messages[0].Message;
-                }
-                Logger.warn("Error fetching tax rate from TaxCloud:", errMsg);
+          try {
+            const response = HTTP.post(url, request);
```

Comment:
Is there another solution rather than making this sync? I feel like I can see a noticeable sluggishness. Since we already have it calculated on the cart I don't quite understand why we need block to calculate this again to get it to show up on the completed order.

Reply:
As you correctly stated, the tax has already been calculated before (through running this very method). The thing to understand here is that this method is always running in conjunction with the method "taxes/calculate" in /imports/plugins/core/taxes/server/methods/methods.js.

So the excecution order is always the following:

The core method "taxes/calculate" sets the tax to 0.00 in Line 245.  After that this method is executed, which overwrites the previously calculated value (0.00) to the value calculated from taxcloud.

If this is method is not synchronous, the 0.00 value from core will be in place when the copyCartToOrder method runs, because the taxcloud value has not arrived yet.

Therefore it's made synchronous (as it is in avalara as well)